cmd/kubectl-gadget/deploy.go

```diff
@@ -362,35 +196,70 @@ func runDeploy(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("invalid argument %q for --hook-mode=[auto,crio,podinformer,nri,fanotify]", hookMode)
 	}
 
-	t, err := template.New("deploy.yaml").Parse(deployYamlTmpl)
+	objects, err := parseK8sYaml(resources.GadgetDeployment)
 	if err != nil {
-		return fmt.Errorf("failed to parse template: %w", err)
+		return err
```

Comment:
I am not sure about setting the parameters on the objects instead of the yaml text before it is parsed. It will make it more difficult when we want to use `kubectl gadget deploy --dry-run` to output the yaml with the substituted `image:` and other parameters, so that users can see what they are applying before doing so.

I would like `.github/workflows/inspektor-gadget.yml` to add a release artifact so users or third party software can install it via:
```
$ kubectl apply -f https://github.com/kinvolk/inspektor-gadget/releases/download/v0.7.0/gadget-deployment.yaml
```
So that it would be possible to write a Headlamp plugin to install Inspektor Gadget without requiring the dependency on the kubectl-gadget binary.

And to make it possible to write a Gardener extension (see [example of Helm chart for a Gardener extension](https://github.com/gardener/gardener-extension-networking-calico/blob/master/charts/gardener-extension-networking-calico/templates/deployment.yaml)). We don't have to prepare Helm charts in this PR of course, but I'm wondering if this PR is making it more difficult.

What do you think?

Reply:
> > I am not sure about setting the parameters on the objects instead of the yaml text before it is parsed.
> 
> I was pragmatic and it is easier to to set the parameters on the objects instead of dealing with strings.
> I can indeed convert the `string` to sort of YAML struct then set the parameters and convert it again to string but I find this operation heavy.

The double conversion would lose the comments too.

I was suggesting to apply the parameters on the Go template like it was done before, and then convert the yaml text with the parameters to objects. In that way, we can have the yaml text with the comments and with the parameters.

> > It will make it more difficult when we want to use kubectl gadget deploy --dry-run to output the yaml with the substituted image: and other parameters, so that users can see what they are applying before doing so.
> 
> I do not understand this point.
> There is no `--dry-run` option and the `kubectl gadget deploy` does not print anything with this modification.
> Can you please clarify it?

I was suggesting that this PR should implement this `--dry-run` option. Today, users can look at the yaml to see what they are about to deploy. With this PR, users can no longer do that. So adding `--dry-run` option would resolve this.

> > I would like .github/workflows/inspektor-gadget.yml to add a release artifact so users or third party software can install it via:
> 
> I think this the goal of `pkg/resources/manifests/deploy.yaml`, people can just apply it and there a ready to play with Inspektor Gadget.

But this file has:
```
        image: "ghcr.io/kinvolk/gadget"
```
without the version attached to it. So users will have to edit the file to write the version they want to use, and this is not documented. If the yaml was attached as a release artifact, they can just use it with `kubectl apply -f https://...` without downloading it and fixing it first.

> Note that this file only deals with k8s resources related to Inspektor Gadget and does not deal with trace CRD.

Understood.

Reply:
> I am really not sure to understand.
> Do you want to keep the `| kubectl apply -f -` to deploy Inspektor Gadget?
> I really want to get rid of it and I think getting rid of it would mean to remove the Go template.

I don't want to keep the `| kubectl apply -f -` but to have an option for users to double check what is going to be deployed, if they want to. I imagine that on production systems, one would want to inspect the yaml. I would like to be able to:
1. `kubectl gadget deploy --print-only` (or `--dry-run`, naming to be decided) to print the yaml on stdout so I can check if it looks ok.
2. `kubectl gadget deploy` to install Inspektor Gadget on Kubernetes.

Before this PR, it was possible. I just don't want to lose this feature.

> As a consequence, people who want to install Inspektor Gadget with only YAML files will need to do the following:
> 
> ```bash
> # Deploy everything related to k8s
> $ kubectl apply -f .../deploy.yaml
> # Deploy everything related to trace CRD
> $ kubectl apply -f .../gadget.kinvolk.io_traces.yaml
> ```
> 
> And `gadget.kinvolk.io_traces.yaml` is not published as release artifact somewhere.

Oh, I didn't understand actually. I guess gadget.kinvolk.io_traces.yaml could be added in the release artifact? But it does not have parameters, so not sure if it's necessary.

Reply:
> without the version attached to it. So users will have to edit the file to write the version they want to use, and this is not documented. If the yaml was attached as a release artifact, they can just use it with kubectl apply -f https://... without downloading it and fixing it first.

What about updating those files before tagging a release to make sure they point to the right image?, then those files could be used directly to install the right version:

```
$ kubectl apply -f https://raw.githubusercontent.com/kinvolk/inspektor-gadget/main/manifests/gadget.yaml
--> installs latest version

$ kubectl apply -f https://raw.githubusercontent.com/kinvolk/inspektor-gadget/v0.7.0/manifests/gadget.yaml
--> installs v0.7.0

$ cat https://raw.githubusercontent.com/kinvolk/inspektor-gadget/v0.7.0/manifests/gadget.yaml | grep image 
        image: "ghcr.io/kinvolk/inspektor-gadget:v0.7.0"
```

Reply:
> What about updating those files before tagging a release to make sure they point to the right image?, then those files could be used directly to install the right version:

It's a bit more complicated to make a release because we can't just git-tag the last commit on `main` but we need to create a commit with the version and then revert it... Do you suggest to create those commits manually, with a script or with a GitHub Action somehow?